### ISSUE TITLE
[SLPVectorizer] Assertion `Ptr != End && "dereferencing end() iterator"' failed

### ISSUE THREAD

**Issue by dcb314** [llvm/llvm-project#85396]
On x86_64 in the last few days, I see a new assertion fail:

[mfx_h264_enc_common_hw-4ee304.sh.gz](https://github.com/llvm/llvm-project/files/14615865/mfx_h264_enc_common_hw-4ee304.sh.gz)
[mfx_h264_enc_common_hw-4ee304.cpp.gz](https://github.com/llvm/llvm-project/files/14615867/mfx_h264_enc_common_hw-4ee304.cpp.gz)

**Comment by nikic** [comment]
Reduced for `-passes=slp-vectorizer`:
```llvm
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-linux-gnu"

define i32 @_ZN15MfxHwH264Encode24CheckVideoParamQueryLikeERNS_13MfxVideoParamERK15MFX_ENCODE_CAPS10eMFXHWType10eMFXVAType12eMFXGTConfig(ptr %hwCaps) #0 {
entry:
  br label %for.cond.cleanup4338

for.cond.cleanup4338:                             ; preds = %entry
  br i1 false, label %if.then4896, label %land.rhs.lr.ph

if.then4896:                                      ; preds = %for.cond.cleanup4338
  br i1 false, label %if.end4915, label %if.then4902

if.then4902:                                      ; preds = %if.then4896
  %MaxNum_WeightedPredL0 = getelementptr i8, ptr %hwCaps, i64 44
  %0 = load i8, ptr %MaxNum_WeightedPredL0, align 4
  %1 = call i8 @llvm.umin.i8(i8 %0, i8 0)
  %.sroa.speculated8478 = zext i8 %1 to i16
  %MaxNum_WeightedPredL1 = getelementptr i8, ptr %hwCaps, i64 45
  %2 = load i8, ptr %MaxNum_WeightedPredL1, align 1
  %3 = call i8 @llvm.umin.i8(i8 %2, i8 0)
  %.sroa.speculated8475 = zext i8 %3 to i16
  br label %if.end4915

if.end4915:                                       ; preds = %if.then4902, %if.then4896
  %maxLuma.sroa.7.0 = phi i16 [ 0, %if.then4896 ], [ %.sroa.speculated8475, %if.then4902 ]
  %maxLuma.sroa.0.0 = phi i16 [ 0, %if.then4896 ], [ %.sroa.speculated8478, %if.then4902 ]
  br label %land.rhs.lr.ph

if.then4921:                                      ; No predecessors!
  br label %land.rhs.lr.ph

land.rhs.lr.ph:                                   ; preds = %if.then4921, %if.end4915, %for.cond.cleanup4338
  %maxLuma.sroa.7.1 = phi i16 [ %maxLuma.sroa.7.0, %if.end4915 ], [ %maxLuma.sroa.7.0, %if.then4921 ], [ 0, %for.cond.cleanup4338 ]
  %maxLuma.sroa.0.1 = phi i16 [ %maxLuma.sroa.0.0, %if.end4915 ], [ %maxLuma.sroa.0.0, %if.then4921 ], [ 0, %for.cond.cleanup4338 ]
  %4 = zext i16 %maxLuma.sroa.0.1 to i64
  %cmp50119196.not.1 = icmp eq i16 %maxLuma.sroa.7.1, 0
  ret i32 0
}

; Function Attrs: nocallback nofree nosync nounwind speculatable willreturn memory(none)
declare i8 @llvm.umin.i8(i8, i8) #1

; uselistorder directives
uselistorder ptr @llvm.umin.i8, { 1, 0 }

attributes #0 = { "target-cpu"="znver3" }
attributes #1 = { nocallback nofree nosync nounwind speculatable willreturn memory(none) }
```
@alexey-bataev

**Comment by alexey-bataev** [comment]
Thanks, helps to investigate corner cases in the new patches

**Comment by alexey-bataev** [comment]
Report was fixed, da118c93b40f74f6770cf8550903721555d3c97b
